third_party/move/move-compiler-v2/src/pipeline/ability_checker.rs

```diff
@@ -78,10 +78,31 @@ fn check_drop_for_temp_with_msg(
     check_drop(func_target, ty, loc, err_msg)
 }
 
+/// If temporary variable `t` does not have ability `drop`, generates an error,
+/// unless the state before `code_offset` won't return.
+fn cond_check_drop_for_temp_with_msg(
+    func_target: &FunctionTarget,
+    code_offset: CodeOffset,
+    t: TempIndex,
+    loc: &Loc,
+    err_msg: &str,
+) {
+    let abort_state = get_abort_state_at(func_target, code_offset);
+    if abort_state.before.may_return() {
+        check_drop_for_temp_with_msg(func_target, t, loc, err_msg)
+    }
+}
+
 /// `t` is the local containing the reference being written to
-fn check_write_ref(target: &FunctionTarget, t: TempIndex, loc: &Loc) {
+/// `code_offset` is the code offset of the WriteRef instruction.
+/// Drop ability must only be enforced if there is an execution path from this code offset
+/// that returns. It's legit in Move to do not drop before an abort or infinite loop.
+fn check_write_ref(target: &FunctionTarget, code_offset: CodeOffset, t: TempIndex, loc: &Loc) {
     if let Type::Reference(_, ty) = target.get_local_type(t) {
-        check_drop(target, ty, loc, "write_ref: cannot drop")
+        let abort_state = get_abort_state_at(target, code_offset);
+        if abort_state.before.may_return() {
```

Comment:
From reading the function's description above, it seems like we could use the after state too. Why not?

I ask because of weird issues that can come up when depending on `before` annotations as is for backwards (and `after` for forwards) analyses (see post analysis fix for live var as well documented example).

Reply:
done